### DISCUSSION TITLE
For Each Child variables: should references be used?

### DISCUSSION THREAD

**Discussion by arthuro555** [4ian/GDevelop#2572] [Enhancement ideas (and their technical discussions)]
As of now, the implementation of the For Each Child variables is not very cool: It uses references for Collections, which is confusing, since the concept of references is nowhere else in GDevelop and not  there for Primitives variables. We should probably replace it with the new deep copying method added with the Arrays and Booleans types, but now I am thinking: Shouldn't we use references there?

I think for other things, users would be confused, but when iterating over the children, in my mind at least, I am still accessing a child of the main variable, not a copy of it.

**Comment by Silver-Streak** [comment]
I can see both sides of this, but you might want to give more detail about what you mean by references so more people can contribute.

Also, doesn't the Raycast conditions use references, even if its just storing them as variables temporarily?

**Reply by arthuro555** [reply]
A reference is a variable that isn't a real variables and redirects all read write operations to the real one.
For example, if `hello` is a reference to `MyStructure.Hey`, then if I modify `MyStructure.Hey`, as `hello` redirects reads it would return `MyStructure.Hey`s new value, and if I write a new value to `hello`, it redirects the write to `MyStructure.Hey` and `MyStructure.Hey` is the one that gets modified. 

It can be confusing as some users might not understand why one variable is changing even if they didn't touch it, because they did touch a variable that is actually a reference. 4ian said multiple times "We do not want references with variables, it'll just confuse users and we don't want to have to explain yet another complex thing to keep GDevelop user friendly", but I think this maybe should be an exception.

**Reply by Silver-Streak** [reply]
Is there anything functionally that _requires_ this to be a reference? 

Because if not, I can see 4ian's logic and would agree.

**Reply by arthuro555** [reply]
As I explained, it isn't required, but it would be IMO in that specific context more intuitive.

**Comment by 4ian** [comment]
I don't think we really should:
* We're in a For Each event. I would actually expect to work with the "real" variable, rather than a copy. This is what a reference represents, but we don't need to explain this in GDevelop: it's another name for your variable, because you're in a for each so of course you need a name to designate this variable.
* Copying would severely impact performance. If you're iterating on a large array (10000 items) containing structures with each 20 child variables... that's 200000 copies for a single frame. Instead of 10000 iterations with 0 copies, just a "reference" in the sense that "this is the same variable, you're just using another name".
* Finally, if you really want to do a "deep copy", you can do it with the actions. But if we make a deep copy by default, then you don't have a way to do a performant for each with "references".

More generally, a deep copy should rather be a conscious choice, done with the action to do it. But by default, it should not be done.
